---
title: "Guide to generating spectrum catalogs and plots from VCFs in the [mSigAct web server](https://msigact.ai)"
author: "Steve Rozen and Nanhai Jiang"
output: md_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Guide to generating spectrum catalogs and plots from VCFs in the [mSigAct web server](https://msigact.ai)

### Introduction

The function of the tab "Generate spectrum catalogs from VCFs" is,
to generate spectrum catalogs from VCFs and also to plot
the spectra.

### The main input area for uploading VCFs

```{r, echo=FALSE, out.width="60%"}
knitr::include_graphics("VCF-to-spectra.guide-images/Slide1.PNG")
```

### Test and example data

These buttons provide some pre-loaded example data and several test VCFs that
you can download and then re-upload.

```{r, echo=FALSE, out.width="60%"}
knitr::include_graphics("VCF-to-spectra.guide-images/Slide2.PNG")
```

### Select the reference genome

```{r, echo=FALSE, out.width="60%"}
knitr::include_graphics("VCF-to-spectra.guide-images/Slide3.PNG")
```

### Select the genomic region

```{r, echo=FALSE, out.width="60%"}
knitr::include_graphics("VCF-to-spectra.guide-images/Slide4.PNG")
```

### Select the variant caller

```{r, echo=FALSE, out.width="60%"}
knitr::include_graphics("VCF-to-spectra.guide-images/Slide5.PNG")
```

### Select and upload one or more VCFs

VCF formats are only partly standardized; please contact us if mSigAct has
difficulty analyzing your VCF.

```{r, echo=FALSE, out.width="60%"}
knitr::include_graphics("VCF-to-spectra.guide-images/Slide6.PNG")
```

