# Cannot use plan(multicore), otherwise the progress bar for asynchronous
# process will not work properly
future::plan(future::multisession(workers = min(64, future::availableCores())))

#' @import mSigAct
#' @import promises
#' @import ipc
#' @import shiny
#' @import shinydashboard
app_server <- function(input, output, session) {
  # List the first level callModules here
  tryCatch({
    
    # Reactive expression for choosing more signatures
    choose.more.sigs <- reactive({
      if (input.catalog.type() == "SBS96") {
        sig.universe <- colnames(COSMIC.v3.genome.SBS96.sigs)
      } else {
        sig.universe <- 
          colnames(PCAWG7::signature[["genome"]][[input.catalog.type()]])
      }
      more.sigs <- setdiff(sig.universe, input$preselectedSigs)
      return(more.sigs)
    }) 
    
    fut <- NULL
    result_val <- reactiveVal()
    running <- reactiveVal(FALSE)
    interruptor <- ipc::AsyncInterruptor$new()
    
    # Create reactiveValues object
    # and set flag to 0 to prevent errors with adding NULL
    rv <- reactiveValues(downloadFlag = 0)
    
    # Increase the file upload limit to 100MB
    options(shiny.maxRequestSize=100*1024^2)
    
    # When user clicks the action link on Home page, direct user to the relevant tab
    observeEvent(input$linkToGenerateCatalogTab, {
      shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                     selected = "generateCatalogTab")
    })
    
    observeEvent(input$linkToUploadSpectraTab, {
      shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                     selected = "uploadSpectraTab")
    })
    
    # Create an empty list which can be used to store notification ids for
    # generating catalogs later
    ids <- list("error" = character(0), "warning" = character(0),
                "message" = character(0))
    
    # Create an empty list which can be used to store the return value of processing VCFs
    retval <- list()
    
    # Create a variable which can be used to store the uploaded catalog later
    catalog <- NULL
    
    list.of.catalogs <- NA
    
    region <- catalog.path <- NA
    
    input.catalog.type <- reactiveVal(NA)
    
    sig.universe.all <- sig.universe <- NULL
    
    showSBS192Catalog <- TRUE
    
    attribution.results <- FALSE
    
    plotdata <- list(cossim = NULL, spect = NULL, 
                     reconstructed.catalog = NULL,
                     sig.universe = NULL, 
                     best.MAP.exp = NULL,
                     dat = NULL)
    
    dat <- data.frame(name = character(), spectrum = character(), 
                      proposed.aetiology = character())
    
    AdjustNumberOfCores <- 
      getFromNamespace(x = "Adj.mc.cores", ns = "mSigAct")
    
    #######################################################################
    # Functions related to UploadVCFUI(), the first tab
    #######################################################################
    # When user uploads VCF files, then show the action button "Generate catalogs"
    observeEvent(input$vcf.files, {
      output$downloadZipFile  <- renderUI({
        MyDownloadButton(
          outputId = "download",
          label = "Create catalogs",
          style="color: #fff;
                   background-color: #337ab7;
                   border-color: #2e6da4;")
      })
    }) 
    
    # Download sample VCFs when user clicks the button
    output$downloadsampleVCFs <- downloadHandler(
      filename = function() {
        "mSigAct-example-VCFs.zip"
      },
      content = function(file) {
        PrepareExampleVCFs(file)
      })
    
    # Download sample catalogs when user clicks the button
    output$downloadSampleSpectra <- downloadHandler(
      filename = function() {
        "mSigAct-example-spectra.zip"
      },
      content = function(file) {
        PrepareExampleSpectra(file)
      })
    
    # Run analysis on sample Strelka SBS VCFs when user clicks the button
    output$runstrelkasbsvcfs <- downloadHandler(
      filename = function() {
        "mSigAct-test-run-Strelka-VCFs-output.zip"
      },
      content = function(file) {
        results <- RunICAMSOnSampleStrelkaVCFs(session, output, file, ids)
        list.of.catalogs <<- results$counts
        region <<- "genome"
        # When the downloadHandler function runs, increment rv$downloadFlag
        rv$downloadFlag <- rv$downloadFlag + 1
      })
    
    # Run analysis on sample Mutect VCFs when user clicks the button
    output$runmutectvcfs <- downloadHandler(
      filename = function() {
        "mSigAct-test-run-Mutect-VCFs-output.zip"
      },
      content = function(file) {
        results <- RunICAMSOnSampleMutectVCFs(session, output, file, ids)
        list.of.catalogs <<- results$counts
        region <<- "genome"
        # When the downloadHandler function runs, increment rv$downloadFlag
        rv$downloadFlag <- rv$downloadFlag + 1
      })
    
    if (FALSE) {
      output$SBS96plot <- NULL
      output$SBS192plot <- NULL
      output$SBS1536plot <- NULL
      output$DBS78plot <- NULL
      output$DBS136plot <- NULL
      output$DBS144plot <- NULL
      output$IDplot <- NULL
    }
    
    # When user submits VCF to analysis, then the program will try to
    # generate a zip archive based on the input files and parameters
    output$download <- downloadHandler(
      
      filename = function() {
        errors <- CheckInputsForVCF(input)
        ids$error <<- append(ids$error, AddErrorMessage(errors))
        paste0(input$zipfile.name, ".zip")
      },
      
      content = function(file) {
        req(input$variantCaller, input$ref.genome, input$region, input$vcf.files)
        if (input$variantCaller == "unknown") {
          if (is.null(input$mergeSBS)) {
            return()
          }
        }
        region <<- input$region
        result <- ProcessVCFs(input, output, file, ids)
        retval <<- result$retval
        old.error.ids <- ids$error
        ids <<- result$ids
        ids$error <- append(ids$error, old.error.ids)
        
        list.of.catalogs <<- retval$counts
        #density.catalog <- retval$density
        
        # When the downloadHandler function runs, increment rv$downloadFlag
        rv$downloadFlag <- rv$downloadFlag + 1

      })
    
    
    observeEvent(rv$downloadFlag, {
      output$showSpectraFromVCF <- renderUI({
        actionButton(inputId = "showSpectraOfVCF", label = "Show spectra",
                     style= "color: #fff; background-color: #337ab7;
                              border-color: #2e6da4; ")
      })
      
      output$sigAttributionFromVCF <- renderUI({
        actionButton(inputId = "sigAttributionOfVCF", label = "Signature attribution",
                     style= "color: #fff; background-color: #337ab7;
                              border-color: #2e6da4; ")
      })
      
      output$selectSampleFromUploadedVCF <- renderUI(
        {
          sample.names <- colnames(list.of.catalogs[[1]])
          radioButtons(inputId = "sampleNameFromUploadedVCF",
                       label = "Select sample",
                       choices = sample.names,
                       selected = character(0))
        }
      )
      #shinyjs::show(id = "sampleNameFromUploadedVCF")
    }, ignoreInit = TRUE)
    
    observeEvent(input$showSpectraOfVCF, {
      # Must delete the previous plot and widget before hiding them
      output$spectraPlotFromCatalog <- NULL
      output$selectSampleFromUploadedCatalog <- NULL
      
      # Hide the previous plot from uploaded catalog
      shinyjs::hide(id = "spectraPlotFromCatalog")
      shinyjs::hide(id = "selectSampleFromUploadedCatalog ")
      
      shinyjs::show(id = "selectSampleFromUploadedVCF")
      
      shinyjs::show(selector = '#panels li a[data-value=showSpectraTab]')
      if (input.catalog.type() %in% c("SBS96", "SBS192", "DBS78", "ID")) {
        shinyjs::show(selector = '#panels li a[data-value=sigAttributionTab2]')
      }
      shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                     selected = "showSpectraTab")
    })
    
    observeEvent(input$sigAttributionOfVCF, {
      # Must delete the previous plot and widget before hiding them
      output$spectraPlotFromCatalog <- NULL
      output$selectSampleFromUploadedCatalog <- NULL
      
      # Hide the previous plot from uploaded catalog
      shinyjs::hide(id = "spectraPlotFromCatalog")
      shinyjs::hide(id = "selectSampleFromUploadedCatalog ")
      
      shinyjs::show(id = "selectSampleFromUploadedVCF")
      
      shinyjs::show(selector = '#panels li a[data-value=showSpectraTab]')
      shinyjs::show(selector = '#panels li a[data-value=sigAttributionTab2]')
      shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                     selected = "sigAttributionTab2")
    })
    
    #######################################################################
    # Start of functions related to ShowSpectraUI
    #######################################################################
    
    # When user selects the sample from uploaded VCF, show
    # the sample's mutational spectrum
    observeEvent(input$sampleNameFromUploadedVCF, {
      output$spectraPlotFromVCF <- renderUI (
        {
          output$SBS96plot <- renderPlot({
            catSBS96 <-
              list.of.catalogs$catSBS96[, input$sampleNameFromUploadedVCF, drop = FALSE]
            ICAMS::PlotCatalog(catSBS96)
          }, height = 230, width = 800)
          
          
          output$SBS192plot <- renderPlot({
            catSBS192 <-
              list.of.catalogs$catSBS192[, input$sampleNameFromUploadedVCF, drop = FALSE]
            ICAMS::PlotCatalog(catSBS192)
          }, height = 250, width = 800)
          
          if (FALSE) {
            output$SBS12plot <- renderPlot({
              catSBS192 <-
                list.of.catalogs$catSBS192[, input$sampleNameFromUploadedVCF, drop = FALSE]
              ICAMS::PlotCatalog(catSBS192, plot.SBS12 = TRUE)
            }, height = 350, width = 350)
            
          }
           
          output$SBS1536plot <- renderPlot({
            catSBS1536 <-
              list.of.catalogs$catSBS1536[, input$sampleNameFromUploadedVCF, drop = FALSE]
            ICAMS::PlotCatalog(catSBS1536)
          }, height = 800, width = 800)
          
          output$DBS78plot <- renderPlot({
            catDBS78 <-
              list.of.catalogs$catDBS78[, input$sampleNameFromUploadedVCF, drop = FALSE]
            PlotCatalog(catDBS78)
          }, height = 250, width = 800)
          
          output$DBS136plot <- renderPlot({
            catDBS136 <-
              list.of.catalogs$catDBS136[, input$sampleNameFromUploadedVCF, drop = FALSE]
            ICAMS::PlotCatalog(catDBS136)
          }, height = 500, width = 700)
          
          output$DBS144plot <- renderPlot({
            catDBS144 <-
              list.of.catalogs$catDBS144[, input$sampleNameFromUploadedVCF, drop = FALSE]
            ICAMS::PlotCatalog(catDBS144)
          }, height = 350, width = 350)
          
          output$IDplot <- renderPlot({
            catID <-
              list.of.catalogs$catID[, input$sampleNameFromUploadedVCF, drop = FALSE]
            ICAMS::PlotCatalog(catID)
          }, height = 230, width = 800)
          
          tabsetPanel(type = "tabs",
                      tabPanel("SBS96", plotOutput("SBS96plot")),
                      tabPanel("SBS192", plotOutput("SBS192plot")),
                      #tabPanel("SBS12", plotOutput("SBS12plot")),
                      tabPanel("SBS1536", plotOutput("SBS1536plot")),
                      tabPanel("DBS78", plotOutput("DBS78plot")),
                      tabPanel("DBS136", plotOutput("DBS136plot")),
                      tabPanel("DBS144", plotOutput("DBS144plot")),
                      tabPanel("ID", plotOutput("IDplot"))
          )
        })
      shinyjs::show(id = "spectraPlotFromVCF")
    })
    
    # When user selects the sample from uploaded catalog, show
    # the sample's mutational spectrum
    observeEvent(input$selectedSampleFromUploadedCatalog, {
      output$SBS96plot <- renderPlot({
        catSBS96 <-
          list.of.catalogs$catSBS96[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catSBS96)
      }, height = 230, width = 800)
      
      
      output$SBS192plot <- renderPlot({
        catSBS192 <-
          list.of.catalogs$catSBS192[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catSBS192)
      }, height = 250, width = 800)
      
      output$SBS12plot <- renderPlot({
        catSBS192 <-
          list.of.catalogs$catSBS192[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catSBS192, plot.SBS12 = TRUE)
      }, height = 350, width = 350)
      
      output$SBS1536plot <- renderPlot({
        catSBS1536 <-
          list.of.catalogs$catSBS1536[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catSBS1536)
      }, height = 800, width = 800)
      
      output$DBS78plot <- renderPlot({
        catDBS78 <-
          list.of.catalogs$catDBS78[, input$sampleNameFromUploadedVCF, drop = FALSE]
        PlotCatalog(catDBS78)
      }, height = 250)
      
      output$DBS136plot <- renderPlot({
        catDBS136 <-
          list.of.catalogs$catDBS136[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catDBS136)
      }, height = 500, width = 700)
      
      output$DBS144plot <- renderPlot({
        catDBS144 <-
          list.of.catalogs$catDBS144[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catDBS144)
      }, height = 350, width = 350)
      
      output$IDplot <- renderPlot({
        catID <-
          list.of.catalogs$catID[, input$sampleNameFromUploadedVCF, drop = FALSE]
        ICAMS::PlotCatalog(catID)
      }, height = 230, width = 800)
      
      
      if (input.catalog.type() == "SBS96") {
        height <- 230
        width <- 800
      } else if (input.catalog.type() == "SBS192") {
        height <- 250
        width <- 800
      } else if (input.catalog.type() == "SBS1536") {
        height <- 800
        width <- 800
      } else if (input.catalog.type() == "DBS78") {
        height <- 250
        width <- 800
      } else if (input.catalog.type() == "DBS136") {
        height <- 500
        width <- 700
      } else if (input.catalog.type() == "DBS144") {
        height <- 350
        width <- 350
      } else if (input.catalog.type() == "ID") {
        height <- 230
        width <- 800
      } 
      
      output$spectraPlotFromCatalog <- renderUI(
        {
          output$spectrum <- renderPlot(
            {
              PlotCatalog(catalog[, input$selectedSampleFromUploadedCatalog,
                                  drop = FALSE])
            }, width = width, height = height)
          plotOutput(outputId = "spectrum")
        }
      )
      shinyjs::show(id = "spectraPlotFromCatalog")
    })
    
    # After showing the spectra plot from catalog, show an actionButton to redirect
    # to "Signature attribution" tab if the spectra catalog type is supported for
    # signature attribution
    observeEvent(CheckArgumentsForSpectra(), {
      
      req(input$ref.genome2, input$region2)
      
      # Delete the previous spectra plot
      output$spectraPlotFromCatalog <- NULL
      
      if (showSBS192Catalog == FALSE) {
        return()
      }
      
      output$selectSampleFromUploadedCatalog <-
        renderUI(
          {
            # Change the region information for SBS192 catalog
            my.region <- ChangeRegionForSBS192Catalog(input, catalog.path)
            
            catalog <<- ICAMS::ReadCatalog(file = catalog.path,
                                           ref.genome = input$ref.genome2,
                                           region = my.region)
            input.catalog.type(CheckCatalogType(catalog)) 
            
            sample.names <- colnames(catalog)
            
            tagList(
              radioButtons(inputId = "selectedSampleFromUploadedCatalog",
                           label = "Select spectrum to view",
                           choices = sample.names,
                           selected = character(0)),
              
              if (input.catalog.type() %in% c("SBS96", "SBS192", "DBS78", "ID")) {
                actionButton(inputId = "clickToSigAttribution",
                             label = "Signature attribution",
                             style= "color: #fff; background-color: #337ab7;
                              border-color: #2e6da4")
              }
              
            )
          }
        ) # end of renderUI
    })
    
    # When user clicks the action button on Show spectra page, direct user to the relevant tab
    observeEvent(input$clickToSigAttribution, {
      shinyjs::show(selector = '#panels li a[data-value=sigAttributionTab2]')
      shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                     selected = "sigAttributionTab2")
    })
    
    # When user submit new catalog for analysis, remove the previous plots
    observeEvent(input$showSpectraOfCatalog, {
      output$SBS96plot <- NULL
      output$SBS192plot <- NULL
      output$SBS1536plot <- NULL
      output$DBS78plot <- NULL
      output$DBS136plot <- NULL
      output$DBS144plot <- NULL
      output$IDplot <- NULL
    })
    
    ########################################################################
    # Start of functions related to UploadSpectraUI
    ########################################################################
    ShowTwoButtons <- function() {
      output$showSpectraFromCatalog <- renderUI(
        actionButton(inputId = "showSpectraOfCatalog", label = "Show spectra",
                     style="color: #fff; background-color: #337ab7;
                          border-color: #2e6da4;")
      )
      output$sigAttributionFromCatalog <- renderUI(
        actionButton(inputId = "sigAttributionOfCatalog", 
                     label = "Signature attribution",
                     style="color: #fff; background-color: #337ab7;
                          border-color: #2e6da4;"))
    }
    
    observeEvent(input$preloadSBS96Spectra, {
      
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "ref.genome2",
                                      selected = "hg19")
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "region2",
                                      selected = "genome")
      catalog.path <<- system.file("extdata/SBS96-mSigAct-example-spectra.csv", 
                                   package = "mSigAct.server")
      input.catalog.type("SBS96") 
      ShowTwoButtons()
    })
    
    observeEvent(input$preloadSBS192Spectra, {
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "ref.genome2",
                                      selected = "hg19")
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "region2",
                                      selected = "transcript")
      catalog.path <<- system.file("extdata/SBS192-mSigAct-example-spectra.csv", 
                                   package = "mSigAct.server")
      input.catalog.type("SBS192") 
      ShowTwoButtons()
    })
    
    observeEvent(input$preloadDBS78Spectra, {
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "ref.genome2",
                                      selected = "hg19")
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "region2",
                                      selected = "genome")
      catalog.path <<- system.file("extdata/DBS78-mSigAct-example-spectra.csv", 
                                   package = "mSigAct.server")
      input.catalog.type("DBS78") 
      ShowTwoButtons()
    })
    
    observeEvent(input$preloadIDSpectra, {
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "ref.genome2",
                                      selected = "hg19")
      shinyWidgets::updatePickerInput(session = session,
                                      inputId = "region2",
                                      selected = "genome")
      catalog.path <<- system.file("extdata/ID-mSigAct-example-spectra.csv", 
                                   package = "mSigAct.server")
      input.catalog.type("ID")
      ShowTwoButtons()
    })
    
    observeEvent(input$upload.spectra, {
      # catalog.info is a data frame that contains one row for each uploaded file,
      # and four columns "name", "size", "type" and "datapath".
      # "name": The filename provided by the web browser.
      # "size": The size of the uploaded data, in bytes.
      # "type": The MIME type reported by the browser.
      # "datapath": The path to a temp file that contains the data that was uploaded.
      catalog.info <- input$upload.spectra
      catalog.path <<- catalog.info$datapath
      ShowTwoButtons()
    })
    
    # Check the arguments for uploaded spectra
    observeEvent(
      CheckArgumentsForSpectra(),
      {
        if (TwoActionButtonsClicked(input) == FALSE) {
          return()
        } else {
          retval <- CheckInputsForSpectra(input, catalog.path)
          ids$error <<- append(ids$error, AddErrorMessage(retval$error))
          showSBS192Catalog <<- retval$SBS192.check
        }
      }
    )
    
    # When user clicks the action button on "Upload Spectra" page, direct user to
    # the relevant tab
    observeEvent(input$showSpectraOfCatalog, {
      if (TwoActionButtonsClicked(input) == FALSE) {
        return()
      } else if (showSBS192Catalog == FALSE) {
        return()
      } else {
        req(input$ref.genome2, input$region2)
        # Hide the widgets for previous uploaded VCF
        output$spectraPlotFromVCF <- NULL
        output$selectSampleFromUploadedVCF <- NULL
        shinyjs::hide(id = "spectraPlotFromVCF")
        shinyjs::hide(id = "selectSampleFromUploadedVCF")
        
        shinyjs::show(selector = '#panels li a[data-value=showSpectraTab]')
        
        if (input.catalog.type() %in% c("SBS96", "SBS192", "DBS78", "ID")) {
          shinyjs::show(selector = '#panels li a[data-value=sigAttributionTab2]')
        }
        shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                       selected = "showSpectraTab")
      }
    })
    
    observeEvent(input$sigAttributionOfCatalog, {
      
      if (TwoActionButtonsClicked(input) == FALSE) {
        return()
      } else if (showSBS192Catalog == FALSE) {
        return()
      } else {
        req(input$ref.genome2, input$region2)
        region <<- input$region2
        
        # Change the region information for SBS192 catalog
        my.region <- ChangeRegionForSBS192Catalog(input, catalog.path)
        
        catalog <<- 
          ICAMS::ReadCatalog(file = catalog.path, ref.genome = input$ref.genome2,
                             region = my.region)
        
       input.catalog.type(CheckCatalogType(catalog))
        
        if (!input.catalog.type() %in% c("SBS96", "SBS192", "DBS78", "ID")) {
          showNotification(ui = "Error:", 
                           action = paste0("Can only do signature attribution ", 
                                           "for SBS96, SBS192, DBS78 and ID"),
                           type = "error", duration = NULL)
          return()
        }
        # Hide the widgets for previous uploaded VCF
        shinyjs::hide(id = "spectraPlotFromVCF")
        shinyjs::hide(id = "selectSampleFromUploadedVCF")
        
        shinyjs::show(selector = '#panels li a[data-value=showSpectraTab]')
        shinyjs::show(selector = '#panels li a[data-value=sigAttributionTab2]')
        shinydashboard::updateTabItems(session = session, inputId = "panels", 
                                       selected = "sigAttributionTab2")
      }
    })
    
    ########################################################################
    # Start of functions related to SignatureAttributionUI
    ########################################################################
    
    # When user clicks either of the two actionButtons "Show spectra", 
    # "Signature attribution" on UploadSpectraUI page, then update
    # the shiny widgets used for attribution analysis
    CheckArgumentsForSpectra <- reactive({
      list(input$showSpectraOfCatalog, input$sigAttributionOfCatalog)
    })
    
    TwoActionButtonsClicked <- function(input) {
      if(is.null(input$showSpectraOfCatalog) && is.null(input$sigAttributionOfCatalog)){
        return(FALSE)
      }
      if(input$showSpectraOfCatalog == 0 && input$sigAttributionOfCatalog == 0){
        return(FALSE)
      }
      return(TRUE)
    }
    
    observeEvent(
      CheckArgumentsForSpectra(),
      {
        req(input$ref.genome2, input$region2)
        
        # Check whether the SBS192 catalog has the correct "region" parameter
        if (showSBS192Catalog == FALSE) {
          return()
        }
        
        output$selectSampleForAttribution <- renderUI(
          { 
            # Change the region information for SBS192 catalog
            my.region <- ChangeRegionForSBS192Catalog(input, catalog.path)
            catalog <<- ICAMS::ReadCatalog(file = catalog.path,
                                           ref.genome = input$ref.genome2,
                                           region = my.region)
            input.catalog.type(CheckCatalogType(catalog))
            sample.names <- colnames(catalog)
            selectInput(inputId = "selectedSampleForAttribution",
                        label = "Select sample",
                        choices = sample.names)
          }
        )
        
        output$selectCancerType <- renderUI(
          {
            cancer.types <-
              c("Unknown", colnames(CancerTypeToExposureStatData()))
            selectInput(inputId = "selectedCancerType",
                        label = "Select cancer type",
                        choices = cancer.types)
          }
        )
        
        output$uploadedCatalogType <- renderUI(
          {
            p(tags$b("Mutation type: "), input.catalog.type())
          }
        )
        shinyjs::show(id = "uploadedCatalogType")
        
        # Hide the UI element previously generated by uploading VCFs
        shinyjs::hide(id = "chooseCatalogType")
        
        # Hide the previous signature aetiology table
        #output$sigAetiologyTable <- NULL
        shinyjs::hide(id = "sigAetiologyTable")
      })
    
    observeEvent(input$selectedCancerType, {
      if (!input.catalog.type() %in% c("SBS96", "SBS192", "DBS78", "ID")) {
        showNotification(ui = "Error:", 
                         action = paste0("Can only do signature attribution ", 
                                         "for SBS96, SBS192, DBS78 and ID"),
                         type = "error", duration = NULL)
        return()
      }
      
      if (input$selectedCancerType != "Unknown") {
        ShowPreselectedSigs(input, output,input.catalog.type())
        } # end of if statement
    
      # Show the actionButton for user to add more signatures
      output$addSig <- renderUI(
        actionButton(inputId = "addMoreSigs", label = "Add more signatures",
                     style= "color: #fff; background-color: #337ab7;
                              border-color: #2e6da4; ")
      )
    }, ignoreInit = TRUE) # end of observeEvent
    
    
    # When user clicks either of the two actionButtons "Show spectra", 
    # "Signature attribution" on UploadVCFUI page, then update
    # the shiny widgets used for attribution analysis
    CheckArgumentsForVCF <- reactive({
      list(input$showSpectraOfVCF, input$sigAttributionOfVCF)
    })
    
    TwoVCFActionButtonsClicked <- function(input) {
      if(is.null(input$showSpectraOfVCF) && is.null(input$sigAttributionOfVCF)){
        return(FALSE)
      }
      if(input$showSpectraOfVCF == 0 && input$sigAttributionOfVCF == 0){
        return(FALSE)
      }
      return(TRUE)
    }
    
    observeEvent(
      CheckArgumentsForVCF(),
      {
        output$selectSampleForAttribution <- renderUI(
          { 
            sample.names <- colnames(list.of.catalogs[[1]])
            selectInput(inputId = "selectedSampleForAttribution",
                        label = "Select sample",
                        choices = sample.names)
          }
        )
        
        output$selectCancerType <- renderUI(
          {
            cancer.types <-
              c("Unknown", colnames(CancerTypeToExposureStatData()))
            selectInput(inputId = "selectedCancerType",
                        label = "Select cancer type",
                        choices = cancer.types)
          }
        )
        
        output$chooseCatalogType <- renderUI(
          {
            selectInput(inputId = "selectCatalogType", 
                        label = "Select catalog type",
                        choices = c("SBS96", "SBS192", "DBS78", "ID"))
          }
        )
        
        shinyjs::show(id = "chooseCatalogType")
        
        # Hide the UI element previously generated by uploading spectra
        shinyjs::hide(id = "uploadedCatalogType")
        
        # Hide the previous signature aetiology table
        #output$sigAetiologyTable <- NULL
        shinyjs::hide(id = "sigAetiologyTable")
      }) # end of observeEvent
    
    CheckArgumentsForAttribution <- reactive({
      list(input$selectedCancerType, input$selectCatalogType)
    })
    
    observeEvent(CheckArgumentsForAttribution(), {
      req(input$selectedCancerType, input$selectCatalogType)
      
      input.catalog.type(input$selectCatalogType) 
      if (is.null(catalog)) {
        catalog.name <- paste0("cat", input.catalog.type())
        catalog <<- list.of.catalogs[[catalog.name]]
      }
      
      ShowPreselectedSigs(input, output, input.catalog.type())
      
      # Show the actionButton for user to add more signatures
      output$addSig <- renderUI(
        actionButton(inputId = "addMoreSigs", label = "Add more signatures",
                     style= "color: #fff; background-color: #337ab7;
                              border-color: #2e6da4; ")
      )
      
    }, ignoreInit = TRUE) # end of observeEvent
    
    # Update the catalog used for attribution if user selects another catalog type
    # for analysis
    observeEvent(input$selectCatalogType, {
      catalog.name <- paste0("cat", input$selectCatalogType)
      catalog <<- list.of.catalogs[[catalog.name]]
    })
    
    observeEvent(input$addMoreSigs, {
      output$chooseMoreSigs <- renderUI(
        {
          shinyWidgets::pickerInput(inputId = "selectedMoreSigs",
                                    label = "Choose more signatures",  
                                    choices = choose.more.sigs(),
                                    options = shinyWidgets::pickerOptions(
                                      actionsBox = TRUE,
                                      dropupAuto = FALSE,
                                      `live-search`=TRUE
                                    ), 
                                    multiple = TRUE)
        }
      )
    })
    
    # Create a reactive expression to determine the signatures used for
    # attribution and then show the signatures aetiology information table
    sigsForAttribution <- reactive({
      list(input$selectCatalogType ,input$preselectedSigs, input$selectedMoreSigs)
      return(c(input$preselectedSigs, input$selectedMoreSigs))
    })
    
    observeEvent(sigsForAttribution(), {
      
      if (is.na(input.catalog.type())) {
        return()
      }
      
      if (input.catalog.type() == "SBS96") {
        sigs.in.correct.order <- 
          intersect(rownames(COSMIC.v3.SBS96.sig.links), sigsForAttribution())
      } else if (input.catalog.type() == "SBS192") {
        sigs.in.correct.order <- 
          intersect(rownames(COSMIC.v3.SBS192.sig.links), sigsForAttribution())
      } else if (input.catalog.type() == "DBS78") {
        sigs.in.correct.order <- 
          intersect(rownames(COSMIC.v3.DBS78.sig.links), sigsForAttribution())
      } else if (input.catalog.type() == "ID") {
        sigs.in.correct.order <- 
          intersect(rownames(COSMIC.v3.ID.sig.links), sigsForAttribution())
      } else {
        sigs.in.correct.order <- NULL
      }
      
      dat <<- PrepareSigsAetiologyTable(input.catalog.type())
      plotdata$dat <<- dat[sigs.in.correct.order, ]
      output$sigAetiologyTable <- DT::renderDataTable({
        DT::datatable(dat[sigs.in.correct.order, ], escape = FALSE, rownames = FALSE,
                      colnames = c("Name", "Signature profile", "Proposed aetiology"), 
                      options = list(lengthMenu = c(25, 50, 75), 
                                     pageLength = 25)) 
      })
      
      if (is.null(sigsForAttribution())) {
        shinyjs::hide(id = "sigAetiologyTable")
      } else {
        shinyjs::show(id = "sigAetiologyTable")
      }
      
      # Don't show "Analyze" button if no signatures were selected in the beginning
      if (!is.null(sigsForAttribution())) {
        output$analysisButton <- renderUI({
          actionButton(inputId = "startAnalysis", label = "Analyze",
                       style= "color: #fff; background-color: #337ab7;
                              border-color: #2e6da4; ")
        })
      } # Need to use ignoreNULL here otherwise sigAetiologyTable will not change
        # if user deselect all signatures
    }, ignoreNULL = FALSE,  ignoreInit = TRUE)
    
    # When user changes input for sample selected and catalog type
    # hide the previous signatuer aetiology table
    hideAetiologyTable <- reactive(
      list(input$selectedSampleForAttribution,
           input$selectCatalogType)
    )
    
    # Asynchronous programming starts from here
    observeEvent(input$startAnalysis, {
      
      if(length(sigsForAttribution()) == 0) {
        showNotification(ui = "Error:", 
                         action = "No signatures selected for attribution analysis",
                         type = "error", duration = NULL)
        return()
      }
      
      # Don't do anything if in the middle of a run
      if(running())
        return(NULL)
      
      running(TRUE)
      spect <- catalog[, input$selectedSampleForAttribution, drop = FALSE]
      catalog.type <- input.catalog.type()
      cancer.type <- input$selectedCancerType
      
      if (catalog.type == "SBS192") {
        sig.universe <<- 
          PCAWG7::signature[["genome"]][[catalog.type]][, sigsForAttribution(), drop = FALSE]
      } else if (catalog.type == "SBS96") {
        sig.universe <<- 
          COSMIC.v3.genome.SBS96.sigs[, sigsForAttribution(), drop = FALSE]
      } else {
        sig.universe <<- 
          PCAWG7::signature[[region]][[catalog.type]][, sigsForAttribution(), drop = FALSE]
      }
      
      # Do the first round of cut-off if there are many signatures in the beginning
      if (ncol(sig.universe) > 15) {
        retval <- 
          mSigAct::OptimizeExposureQPBootstrap(spectrum = spect,
                                               signatures = sig.universe, 
                                               mc.cores = AdjustNumberOfCores(50))
        sig.universe <<- sig.universe[, names(retval$exposure), drop = FALSE]
      }
      
      sigs.prop <- mSigAct::ExposureProportions(
        mutation.type = catalog.type,
        cancer.type   = cancer.type,
        all.sigs      = sig.universe,
        must.include  = colnames(sig.universe))
      
      # Create a Progress object
      progress <- ipc::AsyncProgress$new(session, min = 0, max = 1,
                                         message = "Analysis in progress",
                                         detail = "This may take a while...")
      result_val(NULL)
      fut <- future::future(
        {
          # Close the progress when this reactive exits (even if there's an error)
          # on.exit(progress$close())
          
          # Create a callback function to update progress. Each time this is called, it
          # will increase the progress by that value and update the detail
          updateProgress <- function(value = NULL, detail = NULL) {
            progress$inc(amount = value, detail = detail)
            interruptor$execInterrupts()
          }
          
          # In some cases the first phase of the search, which is a 
          # global search, might find different optima depending
          # on the random seed, so the purpose of set.seed with L'Ecuyer-CMRG
          # is to get reproducible results across mclapply / mcparallel 
          # (https://stat.ethz.ch/R-manual/R-devel/library/parallel/html/mcparallel.html).
          set.seed(102119, kind = "L'Ecuyer-CMRG")
          
          retval <- mSigAct::MAPAssignActivity1(
            spect = spect,
            sigs = sig.universe,
            sigs.presence.prop = sigs.prop,
            max.level = length(sigs.prop) - 1,
            p.thresh = 0.01,
            m.opts = mSigAct::DefaultManyOpts(),
            max.mc.cores = AdjustNumberOfCores(50),
            progress.monitor = updateProgress
          )
          
          return(retval)
        }, seed = TRUE) %...>% {
          retval <- .
          
          plotdata$retval <<- retval
          
          if (retval$success == FALSE || is.null(retval$success)) {
            showNotification(ui = "Message:", 
                             action = retval$messages,
                             type = "message", duration = NULL)
            return()
          } else {
            MAP.best.exp <- retval$MAP
            
            reconstructed.catalog0 <- retval$MAP.recon
            
            cossim <- round(mSigAct::cossim(spect, reconstructed.catalog0), 5)
            
            colnames(reconstructed.catalog0) <- 
              paste0("reconstructed (cosine similarity = ", cossim, ")")
            reconstructed.catalog <- round(reconstructed.catalog0)
            
            plotdata$cossim <<- cossim
            plotdata$spect <<- spect
            plotdata$reconstructed.catalog <<- reconstructed.catalog
            plotdata$sig.universe <<- sig.universe
            plotdata$best.MAP.exp <<- MAP.best.exp
            
            retval <- PrepareAttributionResults2(input, output, session, 
                                                 input.catalog.type(), 
                                                 plotdata)
            attribution.results <<- retval$attribution.results
          }
        } %...>% result_val
      
      # Show notification on error or user interrupt
      fut <- promises::catch(fut,
                             function(e){
                               result_val(NULL)
                               print(e$message)
                               showNotification(e$message, 
                                                duration = NULL)
                             })
      
      # When done with analysis, remove progress bar
      fut <- promises::finally(fut, function(){
        progress$close()
        running(FALSE) # Declare done with run
      })
      
      # Return something other than the future so we don't block the UI
      NULL
    })
    
    # Send interrupt signal to future
    observeEvent(input$cancel,{
      if(running()) {
        interruptor$interrupt("Task cancelled")
      }
    })
    
  }) # end of tryCatch
}
  
